rules/Php56/NodeAnalyzer/UndefinedVariableResolver.php

```diff
@@ -142,19 +142,19 @@ private function isAssign(Node $parentNode): bool
 
     private function shouldSkipVariable(Variable $variable, Node $parentNode): bool
     {
-        if ($this->variableAnalyzer->isStaticOrGlobal($variable)) {
+        if ($this->isAsCoalesceLeftOrAssignOpCoalesceVar($parentNode, $variable)) {
             return true;
         }
 
-        if ($this->isAssign($parentNode)) {
+        if ($this->variableAnalyzer->isStaticOrGlobal($variable)) {
             return true;
         }
 
-        if ($this->issetOrUnsetOrEmptyParent($parentNode)) {
+        if ($this->isAssign($parentNode)) {
             return true;
         }
 
-        if ($this->isAsCoalesceLeftOrAssignOpCoalesceVar($parentNode, $variable)) {
+        if ($this->issetOrUnsetOrEmptyParent($parentNode)) {
```

Comment:
@samsonasik I think it could make sense to re-order the skipchecking methods so e.g. doing AST traversing only after cheaper checks have been done

Reply:
This patch is to ensure no infinite loop by check coalesce early, which check coalesce later cause infinite loop, see https://getrector.com/demo/289e4c81-adae-4062-8a7a-2ad51c53c711 

feel free to provide improvement PR 👍

Reply:
@staabm I created PR https://github.com/rectorphp/rector-src/pull/3643 for that.